### ISSUE TITLE
Childframe covers company popup

### ISSUE THREAD

**Issue by casouri** [casouri/eldoc-box#15]
From #11

**Comment by casouri** [comment]
@andreyorst I made a patch and it seems to work for the simplest case. Let's use it for a while and see if other problems occur.

**Comment by andreyorst** [comment]
great! will it be available on melpa or should I install it from git and use some branch to test?

**Comment by casouri** [comment]
Melpa is up to date, you can just use that. 😄

**Comment by andreyorst** [comment]
Thanks, I've tried that, I think it's nice workaround, but not as final solution, since it simply displays eldoc as it would without your package, and I've installed your package to stop that behavior, because it visually moves my buffers when shown or hidden. Another possible workaround would be to display it not at point, like when `eldoc-box-hover-at-point-mode` is disabled, which would be nicer, since it still uses childframes, but positions those in non distracting way.

I wonder I you can borrow some ideas from https://github.com/expez/company-quickhelp to display eldoc at point, but position it correctly accordingly to company.

**Comment by casouri** [comment]
Sounds like a good idea. You can try the patch here: https://github.com/casouri/eldoc-box/tree/cover-company-fix

However, this would cover company-quickhelp if it is enabled, what should we do in that case? Judging from the source of company-quickhelp, there's no good way to get the geometry of the pos-tip it uses.

**Comment by andreyorst** [comment]
> However, this would cover company-quickhelp if it is enabled, what should we do in that case? Judging from the source of company-quickhelp, there's no good way to get the geometry of the pos-tip it uses.

I guess in this way eldoc-box should disable at-point hovering and fallback to posframe at default position. 

> Sounds like a good idea. You can try the patch here: https://github.com/casouri/eldoc-box/tree/cover-company-fix

I'm getting these errors:

```
error in process filter: eldoc-box--default-at-point-position-function-1: Symbol’s value as variable is void: pos-in-window
error in process filter: Symbol’s value as variable is void: pos-in-window
```

**Comment by casouri** [comment]
> I guess in this way eldoc-box should disable at-point hovering and fallback to posframe at default position. 

That's a good idea. Should we just let users do that? I.e., don't use at-point-hover when you also use company-quickhelp.

> I'm getting these errors:

I don't get these errors. And my patch didn't touch that part of the code. Could you verify if you are at f292d28712b38c4c94e5bb9fcf3788f19b83dc6d?

**Comment by andreyorst** [comment]
>That's a good idea. Should we just let users do that? I.e., don't use at-point-hover when you also use company-quickhelp.

yes, I guess.

>I don't get these errors. And my patch didn't touch that part of the code. Could you verify if you are at f292d28?

I've just switched branch and copied file contents, since I don't know how to install packages from git with use-package

**Comment by casouri** [comment]
That's strange, I cloned the branch fresh and loaded the file, and still don't see the error.

**Comment by andreyorst** [comment]
yes, If I clone and load it works fine.

**Comment by andreyorst** [comment]
Tried your package again with Mate (still waiting when Gnome Shell related issue will be fixed) and I think this issue is fixed and can be closed. Eldoc Box shows up correctly near the company popup and follows cursor as you type.

However I have another thoughts about eldoc-box behavior, and I'll describe those in another issue. Thanks!

**Comment by casouri** [comment]
Cool

**Comment by andreyorst** [comment]
https://github.com/casouri/eldoc-box/issues/21#issuecomment-537955462

Ah, finally figured out how to reproduce this issue:
Here I change the size of treemacs buffer. If you toggle these two images you'll notice that `eldoc-box` position is not changed between those [[1]](https://user-images.githubusercontent.com/19470159/66563417-68dcc080-eb66-11e9-95a3-4889fc4fd7d7.png), [[2]](https://user-images.githubusercontent.com/19470159/66563487-8742bc00-eb66-11e9-971f-11a0016959ba.png)

`display-line-number-mode` also doesn't affect `eldoc-box` position: [[1]](https://user-images.githubusercontent.com/19470159/66563248-11d6eb80-eb66-11e9-9b8a-199573ba991a.png), [[2]](https://user-images.githubusercontent.com/19470159/66563339-3fbc3000-eb66-11e9-9f22-99cac0f393dc.png)

Disabling `treemacs` and `display-line-numbers-mode` fixes the issue. 
Actually any left window breaks positioning in right window for some reason:

- `imenu-list` buffer affects position of `eldoc-box`:
  ![image](https://user-images.githubusercontent.com/19470159/66562996-7fcee300-eb65-11e9-994a-73aa2b35bf1e.png)

- Buffer on the left affects position of `eldoc-box` in right buffer:
  ![image](https://user-images.githubusercontent.com/19470159/66562915-501fdb00-eb65-11e9-9dd0-138ab4a7c880.png)

@casouri can you reopen this issue?

**Comment by andreyorst** [comment]
Can be reproduced without any configuration in `emacs -q` after evaluating

```emacs-lisp
(progn (package-initialize) (require 'eldoc-box) (require 'eglot) (require 'company))
```

Enabling `eglot`, `company-mode`, `eldoc-box-hover-at-point-mode` and `display-line-numbers-mode` produces this result:

![image](https://user-images.githubusercontent.com/19470159/66563990-a857dc80-eb67-11e9-826c-f824219e44fd.png)

**Comment by casouri** [comment]
Thanks for your work. I’ll look into it.

**Comment by casouri** [comment]
The line-numbers-mode problem should go away now.

**Comment by casouri** [comment]
Since you use Linux, could you tell me whether the y-position is off when you turn on tool-bar-mode?

**Comment by andreyorst** [comment]
@casouri I can't check, since I'm also experiencing this issue #24

**Comment by andreyorst** [comment]
Changing code like this:

```diff
-  (let* ((pos (pos (posn-at-point point window)))
+  (let* ((pos (posn-at-point point window))
          (x-y (posn-x-y pos))
          (window (posn-window pos))
          (edges (window-body-pixel-edges window)))
```

makes it work, but position is not correct:

![image](https://user-images.githubusercontent.com/19470159/68070302-d5219d00-fd7d-11e9-9f17-872c2bed288c.png)
![image](https://user-images.githubusercontent.com/19470159/68070307-e36fb900-fd7d-11e9-9709-2f238c7d1bb3.png)

As for `tool-bar-mode` it affects the position in a correct way. E.g. the position on screenshots above do not change with or without tool bar

**Comment by andreyorst** [comment]
Yet it seems that this positioning problem is shared among other plugins like company-box. So this may be GNOME Shell related.

**Comment by casouri** [comment]
Ok, here's what I know. `display-line-numbers-mode` invoked problem because it is squeezed into the window, therefore, the x value returned by company is off the width of the line numbers. My last fix fixes that problem (for me). I'm not sure why `imenu-list` has a problem - it seems to use a regular window and it worked fine for me. Maybe you mean the childframe didn't automatically update position when imenu-list popup up?

**Comment by andreyorst** [comment]
It seems that the problem with the leftmost windows is fixed as well

I just mean that instead of displaying under the cursor, box is displayed over the current line

![image](https://user-images.githubusercontent.com/19470159/68071979-69e1c600-fd91-11e9-990d-9eb45d58bbcd.png)
(cursor is after `args|`)

**Comment by casouri** [comment]
Yes, that's what I'm trying to solve. That happens when I endable `tool-bar-mode`. You enabled `tool-bar-mode`, no?

**Comment by andreyorst** [comment]
Tool bar mode doesn't affect position in my case.

**Comment by andreyorst** [comment]
Found it. `global-tab-line` mode is the root of the issue. It's the same as `header-line` but uses additional `tab-line` area added in Emacs 27

**Comment by casouri** [comment]
I see, let me try to figure out how to get the height of the tab-line.

**Comment by casouri** [comment]
I think I got it for the most part. `gloabl-tab-bar-line` doesn't create problem now at least. The positions in a window/frame in Emacs is such a mess...

**Comment by casouri** [comment]
BTW, could you check if the childframe is off by 1 pixel? I observed it on my machine:

<img width="276" alt="Screen Shot 2019-11-02 at 10 37 22 AM" src="https://user-images.githubusercontent.com/13118647/68072489-c2e33700-fd5c-11e9-9ee8-95dad10228c5.png">

**Comment by andreyorst** [comment]
> I think I got it for the most part. `gloabl-tab-bar-line` doesn't create problem now at least. The positions in a window/frame in Emacs is such a mess...

You can look at [this library](https://github.com/tjarvstrand/pos-tip). It has no problems with resize in Gnome and positioning as far as I can tell.


> BTW, could you check if the childframe is off by 1 pixel? I observed it on my machine:
> 
> <img alt="Screen Shot 2019-11-02 at 10 37 22 AM" width="276" src="https://user-images.githubusercontent.com/13118647/68072489-c2e33700-fd5c-11e9-9ee8-95dad10228c5.png">
> 
> Well maybe that's not exactly 1 pixel, but you know what I mean.

![image](https://user-images.githubusercontent.com/19470159/68073127-f0040980-fd9d-11e9-8de6-16318148b9a0.png)

It seems there's no such problem on my machine.
